File: ios/StrainEase/Services/StrainDirectoryCache.swift
import FirebaseFirestore
import Foundation

/// Per-type strain directory cache. Reads three Firestore documents
/// (`strainDirectory/byType/{indica,sativa,hybrid}/current`) written by
/// the daily `warmStrainDirectory` Cloud Function and makes them
/// available to `StrainCatalog.merge(_:preferringType:)`. Mirrors the
/// web client (`src/lib/strain-cache.ts`); the bundled `strain-directory.json`
/// stays as the cold-start fallback when Firestore is offline or the doc
/// is missing.
enum StrainDirectoryCache {
    /// In-memory per-type profiles. Empty until `warm()` resolves (or fails).
    private static var byType: [StrainType: [StrainProfile]] = [:]
    private static var isLoaded: Bool = false
    private static var inflight: Task<Void, Never>?

    /// Kick off the per-type fetch. Idempotent — repeated calls reuse the
    /// in-flight Task. Called from `StrainEaseApp.init` so the cache is
    /// usually loaded by the time `StrainCatalog.merge` runs.
    @MainActor
    static func warm() {
        if isLoaded { return }
        if inflight != nil { return }
        inflight = Task {
            await loadAllTypes()
        }
    }

    /// Force-reload (e.g. pull-to-refresh on Discover). Cancels any
    /// in-flight fetch before starting a new one.
    @MainActor
    static func reload() async {
        inflight?.cancel()
        inflight = nil
        isLoaded = false
        byType = [:]
        await loadAllTypes()
    }

    /// Per-type snapshot loaded from Firestore, or nil if the cache is
    /// still loading / failed. Callers should fall back to the bundled
    /// directory when nil.
    static func profiles(for type: StrainType) -> [StrainProfile]? {
        guard isLoaded else { return nil }
        return byType[type] ?? []
    }

    /// True once the per-type cache has resolved at least one document.
    /// Useful for tests and for the Discover view that wants to wait
    /// for the rich snapshot before rendering the empty-state.
    static var loaded: Bool { isLoaded }

    private static func loadAllTypes() async {
        let types: [StrainType] = [.indica, .sativa, .hybrid]
        let fetched: [(StrainType, [StrainProfile])] = await withTaskGroup(
            of: (StrainType, [StrainProfile]).self
        ) { group in
            for type in types {
                group.addTask {
                    let profiles = (try? await fetch(type: type)) ?? []
                    return (type, profiles)
                }
            }
            var collected: [(StrainType, [StrainProfile])] = []
            for await pair in group {
                collected.append(pair)
            }
            return collected
        }
        for (type, profiles) in fetched {
            byType[type] = profiles
        }
        // Loaded means we hit Firestore (even if every doc was missing).
        // Callers fall back to bundled JSON per-type on empty buckets.
        isLoaded = true
    }

    private static func fetch(type: StrainType) async throws -> [StrainProfile] {
        let snap = try await Firestore.firestore()
            .collection("strainDirectory")
            .document("byType")
            .collection(type.rawValue)
            .document("current")
            .getDocument()
        guard snap.exists, let data = snap.data() else { return [] }
        guard
            let previews = data["previews"] as? [[String: Any]],
            let fetchedAt = data["fetchedAt"] as? Int
        else { return [] }
        // 24h freshness rule matches the web client. Stale docs fall
        // back to the bundled JSON on the merge side.
        let ageMs = Int(Date().timeIntervalSince1970 * 1000) - fetchedAt
        if ageMs > 24 * 60 * 60 * 1000 { return [] }
        return previews.compactMap { row in
            guard
                let name = row["name"] as? String,
                !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            else { return nil }
            let rawType = row["type"] as? String ?? ""
            let resolved: StrainType?
            switch rawType {
            case "indica": resolved = .indica
            case "sativa": resolved = .sativa
            case "hybrid": resolved = .hybrid
            default: resolved = nil
            }
            guard let resolved else { return nil }
            return StrainProfile(
                name: name,
                inKnowledgeBase: true,
                type: resolved,
                thcRange: row["thcRange"] as? String,
                medicalUses: row["medicalUses"] as? [String],
                imageUrl: row["imageUrl"] as? String,
                leaflyRating: row["leaflyRating"] as? Double,
                weedmapsRating: row["weedmapsRating"] as? Double
            )
        }
    }
}

File: android/app/src/main/java/ai/strainease/app/data/StrainDirectoryCache.kt
package ai.strainease.app.data

import android.util.Log
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.FirebaseFirestore
import ai.strainease.app.models.StrainProfile
import ai.strainease.app.models.StrainType
import ai.strainease.app.services.FirebaseBootstrap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

/**
 * Per-type strain directory cache. Reads three Firestore documents
 * (`strainDirectory/byType/{indica,sativa,hybrid}/current`) written by
 * the daily `warmStrainDirectory` Cloud Function and publishes them as
 * a [StateFlow] so [StrainCatalog.merge] prefers the Firestore bucket
 * over the bundled `strain-directory.json` snapshot.
 *
 * Mirrors:
 *   - iOS  `ios/StrainEase/Services/StrainDirectoryCache.swift`
 *   - Web  `src/lib/strain-cache.ts` (read) + `src/lib/strain-catalog.ts` (resolve)
 *
 * Initialised once from `StrainEaseApplication.onCreate` via
 * [StrainCatalog.init], after `FirebaseBootstrap.configure` has run.
 * Without Firebase (or before init completes) `profilesFor` returns an
 * empty list and `StrainCatalog.merge` falls back to the bundled JSON.
 */
object StrainDirectoryCache {

    private const val TAG = "StrainDirectoryCache"
    private const val COLLECTION = "strainDirectory"
    private const val BY_TYPE_DOC = "byType"
    private const val SUBDOC = "current"
    private const val TTL_MS = 24L * 60L * 60L * 1000L // 24 hours, matches iOS/web

    /** Firestore type names are lowercase to match the backend
     *  `partitionByType` keys (functions/src/strain-directory-cache.ts). */
    private fun collectionName(type: StrainType): String = when (type) {
        StrainType.Indica -> "indica"
        StrainType.Sativa -> "sativa"
        StrainType.Hybrid -> "hybrid"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _byType = MutableStateFlow<Map<StrainType, List<StrainProfile>>>(emptyMap())
    val byType: StateFlow<Map<StrainType, List<StrainProfile>>> = _byType.asStateFlow()

    private val _loaded = MutableStateFlow(false)
    /** True once the cache has resolved at least one bucket (even if
     *  the Firestore fetch failed — empty buckets are still cached).
     *  Callers fall back to the bundled JSON on empty buckets. */
    val loaded: StateFlow<Boolean> = _loaded.asStateFlow()

    /** Fetch all three per-type buckets in parallel. Idempotent — safe
     *  to call repeatedly; subsequent calls re-fetch. */
    fun warm() {
        if (!FirebaseBootstrap.isConfigured) return
        scope.launch {
            try {
                val results = StrainType.entries.map { type ->
                    async { type to fetchOne(type) }
                }.awaitAll()
                val next = results.toMap()
                _byType.value = next
                _loaded.value = true
            } catch (t: Throwable) {
                Log.w(TAG, "Failed to load strainDirectory/byType: ${t.message}")
            }
        }
    }

    /** Synchronous read of one bucket's latest snapshot. Returns an
     *  empty list when the cache hasn't loaded yet or that bucket was
     *  empty. Callers should fall back to the bundled JSON on empty. */
    fun profilesFor(type: StrainType): List<StrainProfile> =
        _byType.value[type].orEmpty()

    private suspend fun fetchOne(type: StrainType): List<StrainProfile> {
        val fs = try {
            FirebaseFirestore.getInstance(FirebaseApp.getInstance())
        } catch (t: Throwable) {
            Log.w(TAG, "Firestore not available for ${type.name}: ${t.message}")
            return emptyList()
        }
        return try {
            val snap = fs
                .collection(COLLECTION)
                .document(BY_TYPE_DOC)
                .collection(collectionName(type))
                .document(SUBDOC)
                .get()
                .await()
            if (!snap.exists()) return emptyList()
            val previews = snap.get("previews") as? List<*> ?: return emptyList()
            val fetchedAt = (snap.get("fetchedAt") as? Number)?.toLong() ?: 0L
            val ageMs = System.currentTimeMillis() - fetchedAt
            if (ageMs > TTL_MS) return emptyList()
            previews.mapNotNull { row -> toProfile(row) }
        } catch (t: Throwable) {
            Log.w(TAG, "Failed to fetch ${type.name.lowercase()} bucket: ${t.message}")
            emptyList()
        }
    }

    /** Map a Firestore preview row (Map<String, Any?>) into a
     *  [StrainProfile]. The backend only writes rows whose `type` is
     *  one of the three buckets, so the cast is safe; any unparseable
     *  row is dropped instead of leaking null into the typed field. */
    private fun toProfile(row: Any?): StrainProfile? {
        val map = row as? Map<*, *> ?: return null
        val name = map["name"] as? String ?: return null
        if (name.isBlank()) return null
        val rawType = map["type"] as? String ?: return null
        val type = when (rawType.lowercase()) {
            "indica" -> StrainType.Indica
            "sativa" -> StrainType.Sativa
            "hybrid" -> StrainType.Hybrid
            else -> return null
        }
        return StrainProfile(
            name = name,
            inKnowledgeBase = true,
            type = type,
            thcRange = map["thcRange"] as? String,
            medicalUses = (map["medicalUses"] as? List<*>)
                ?.mapNotNull { it as? String }
                ?.takeIf { it.isNotEmpty() },
            imageUrl = map["imageUrl"] as? String,
            leaflyRating = (map["leaflyRating"] as? Number)?.toDouble(),
            weedmapsRating = (map["weedmapsRating"] as? Number)?.toDouble(),
        )
    }
}
